### PR TITLE
URP Terrain: Fix terrain opacity blend and documentation

### PR DESCRIPTION
**Pull request by spaceburp**
### Checklist for PR maker
- [x] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [x] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [x] Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [x] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
Fixes opacity as density blend for <= 4 layers for URP Terrain Shader and disables density blending when Terrain has > 4 Terrain Layers
Adds to HDRP Terrain Shader documentation for the Enable Opacity as Density feature on Terrain Layers
Adds documentation to URP for Terrain Lit Shader, which is mostly a copy of the HDRP minus the HDRP specific features

Fogbugz Link: https://fogbugz.unity3d.com/f/cases/1283124/

---
### Testing status
Describe what manual/automated tests were performed for this PR

- Manually tested painting and verifying that artifacts previously seen have been removed
- Manually tested 2 - 4 layer blending with different layers with different Opacity as Density states
- Manually tested +4 layer blending and verified that density blending is disabled with >4 Terrain Layers
- Manually tested Heightbased Blending feature for < 4 layers to make sure nothing broke there
- Ran automated tests, specifically the one that was originally failing and made us aware of this issue in the first place: 038_Terrain_Blend
- Created automated test for Opacity as Density blend with < 4 layers

After changes:

4 Layers with 1 layer using Opacity As Density blend. Ignore the black pixels that show. This is because the splat is fully weighted towards that Terrain Layer so it is pulling the black from the Diffuse Map and is not a blending error

https://user-images.githubusercontent.com/1635624/107831371-c7a9df80-6d42-11eb-83cf-c54a4483efc9.mp4

Example of Opacity as Density blending being disabled if there is more than 4 Terrain Layers on the Terrain

https://user-images.githubusercontent.com/1635624/107831379-ca0c3980-6d42-11eb-816c-4b840462852f.mp4

Example of 4 Layers with 2 layers using Opacity As Density Blend

https://user-images.githubusercontent.com/1635624/107831381-cb3d6680-6d42-11eb-9ea3-e90401cf7a68.mp4

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Documentation~/Terrain-Lit-Shader.md
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/Documentation~/Terrain-Lit-Shader.md:43]: Indicates whether HDRP uses the alpha channel of the **Diffuse Map** as a threshold value for layer blending. If the splatmap's weight value plus the alpha channel value is greater than one, the Terrain Layer is visible as part of the blended result. After this calculation, the shader normalizes the modified weights.
<br/>
This property only appears when you enable **Enable Height-based Blend** on the Terrain Layer's TerrainLit Material.

Comment (JordanL8) [reply com.unity.render-pipelines.high-definition/Documentation~/Terrain-Lit-Shader.md:43]: ^^ This is my initial thought. One thing that I want to know is the significance of the shader normalizing the modified weights. What are they used for?

Comment (spaztronaut) [reply com.unity.render-pipelines.high-definition/Documentation~/Terrain-Lit-Shader.md]: Reads well to me. Technically speaking, the shader normalizes the weights after the density blend is calculated so that at least one layer is always visible and it tends to be the layer with the greatest authored splatmap weight but maybe that's not necessary to know?

**Review by phi-lira — APPROVED** [review]
We need to add the same test files for Consoles as well before landing this PR. Ping me for help.

**Review by Verasl — APPROVED** [review]
LGTM

File: com.unity.render-pipelines.high-definition/Documentation~/Terrain-Lit-Shader.md
Comment (spaztronaut) [reply com.unity.render-pipelines.high-definition/Documentation~/Terrain-Lit-Shader.md]: going to remove this from the doc. sharlene is adding this info to the Terrain Layer manual page https://docs.unity3d.com/Manual/class-TerrainLayer.html

**Comment by spaztronaut** [comment]
hey @oleks-k adding you for docs review. adding a page for URP terrain lit shader

File: com.unity.render-pipelines.universal/CHANGELOG.md
Comment (phi-lira) [inline com.unity.render-pipelines.universal/CHANGELOG.md:14]: Merge error

**Review by phi-lira — CHANGES_REQUESTED** [review]

**Comment by phi-lira** [comment]
checking status on this PR again. Let me know when reviewers finished fixing. There's a change to our graphics project coming so more likely there will be some conflicts.

File: com.unity.render-pipelines.universal/CHANGELOG.md
Comment (spaztronaut) [reply com.unity.render-pipelines.universal/CHANGELOG.md]: ah thanks

**Review by GrantLamb-Unity — APPROVED** [review]
Verified the fix on this branch with 2021.2.0a1 and it looked good. I also reviewed the test scenes and found no issues.

**Comment by nerites** [comment]
I tested on a side branch that I made these changes in first. I updated a reference image for URP_Terrain on metal, but aside from that, all remaining failures are failures that replicated on master. For transparency, here is a list of failures grouped by type:

**Failures that replicate on master:**
3 image mismatches on GraphicsTests.TestFarAwayFromOrigin, GraphicsTests.TestMBPCompatibility, and GraphicsTests.TestShaderGraphProperties
- HDRP_Hybrid on Win_DX11_playmode_mono_Linear
- Universal_Hybrid on OSX_Metal_playmode_mono_Linear on version trunk
- Universal_Hybrid on Win_DX11_playmode_mono_Linear

Image mismatch on UniversalGraphicsTests.134_Lighting_Mixed_ShadowMask_Distance
- URP_Lighting on iPhone_Metal_Standalone_cache_il2cpp_Linear

com.unity.test-framework.build: Error when executing git command. git@github.cds.internal.unity3d.com: Permission denied (publickey).
- VFX_HDRP on Win_DX12_mono_Linear_Standalone_cache_build_Player
- VFX_HDRP on Win_DX11_mono_Linear_Standalone_cache_build_Player
- VFX_URP on Win_DX11_mono_Linear_Standalone_cache_build_Player

**Failures that passed on rerun:**
- HDRP on OSX_Metal_playmode_cache_mono_Linear
- URP_Foundation on Win_DX12_mono_Linear_Standalone_cache_build_Player

**Comment by nerites** [comment]
It looks like @hdb-unity is on vacation. Does anybody know who I can get a review from in his place?

**Comment by phi-lira** [comment]
Split project has landed, so you have to move the test file changes / new tests to Terrain project 
 It's inside the same github but in a different project.
 https://github.com/Unity-Technologies/Graphics/tree/master/TestProjects/UniversalGraphicsTest_Terrain

**Review by phi-lira — CHANGES_REQUESTED** [review]
Split project landed, test files related to terrain need to be moved to terrain project
 https://unity.slack.com/archives/GBJ7AU4V8/p1613463834341500

**Comment by wyatttttt** [comment]
> Split project landed, test files related to terrain need to be moved to terrain project
> https://unity.slack.com/archives/GBJ7AU4V8/p1613463834341500

ok. i have this done. just need to push i think

**Comment by wyatttttt** [comment]
running tests again

**Comment by wyatttttt** [comment]
fixed failing android test. should be good now with the tests related to this PR. the other URP failures are these tests on OSX and Windows DX11 jobs:
- GraphicsTests.TestShaderGraphProperties
- GraphicsTests.TestMBPCompatibility
- GraphicsTests.TestFarAwayFromOrigin

not sure what the min version tests are and why they are failing

**Review by hdb-unity — APPROVED** [review]
**Note:** No testing done. Did not check CI results.

My review is required just to confirm that UniversalGraphicsTest changes have been moved to a split project (in this case _Terrain).

3x files changed in UniversalGraphicsTest - please make sure they are reverted if no longer needed, or make sure the same changes are also in _Terrain project.

**Comment by wyatttttt** [comment]
> **Note:** No testing done. Did not check CI results.
> 
> My review is required just to confirm that UniversalGraphicsTest changes have been moved to a split project (in this case _Terrain).
> 
> 3x files changed in UniversalGraphicsTest - please make sure they are reverted if no longer needed, or make sure the same changes are also in _Terrain project.

thanks. fixed the changes to the UniversalGraphicsTest project

File: com.unity.render-pipelines.universal/Documentation~/Terrain-Lit-Shader.md
Comment (sharlenet) [inline com.unity.render-pipelines.universal/Documentation~/Terrain-Lit-Shader.md:25]: This is the link to Paint Holes in the main manual: https://docs.unity3d.com/Manual/terrain-PaintHoles.html

File: com.unity.render-pipelines.universal/Documentation~/Terrain-Lit-Shader.md
Comment (sharlenet) [inline com.unity.render-pipelines.universal/Documentation~/Terrain-Lit-Shader.md:20]: Here's the link to Terrain Settings in the main manual, in case you need it: https://docs.unity3d.com/Manual/terrain-OtherSettings.html

**Comment by wyatttttt** [comment]
failing tests for `Universal_Hybrid on Win_DX11_playmode_mono_Linear on version trunk` and `Universal_Hybrid on OSX_Metal_playmode_mono_Linear on version trunk` are:
- GraphicsTests.TestShaderGraphProperties
- GraphicsTests.TestMBPCompatibility
- GraphicsTests.TestFarAwayFromOrigin

**Review by hdb-unity — COMMENTED** [review]
Note: No manual testing performed by myself. Failed CI checks might be related to known failures that affect base (master branch) as mentioned above.

Developer testing + QA pass seems sufficient.

**Review by hdb-unity — APPROVED** [review]

**Review by phi-lira — APPROVED** [review]

**Comment by ginkgoco** [comment]
looks like there been some error introduced into a good chunk of the terrain test images. will investigate

**Pull request merged**
